lib/GcInfo/GcInfo.cpp

```diff
@@ -0,0 +1,271 @@
+
+  // TODO: Determine call-site-size accurately
+  // https://github.com/Microsoft/llvm/issues/56
+  // Call-site size is not yet available in LLVM's StackMap,
+  // so make up a value for now.
+  // The Call-instruction generated by LLILC on X86/X64 is typically
+  // Call [rax], which has a two-byte encoding.
+  uint8_t CallSiteSize = 2;
+
+#if !defined(NDEBUG)
+  if (EmitLogs) {
+    dbgs() << "  FunctionEntry: " << FunctionEntry << "\n"
+           << "  #Safepoints: " << NumCallSites << "\n";
+  }
+#endif // !NDEBUG
+
+  size_t RecordIndex = 0;
+  for (const auto &R : StackMapParser.records()) {
+
+    // InstructionOffset:
+    // + OffsetCorrection: to account for any bytes before the start
+    //                     of the function.
+    // (-1) CLR expects the statepoint to be one byte before the start of the
+    //     instruction following the Call-instruction.
+    // (-1) Safepoints are encoded with a -1 adjustment.
+    unsigned InstructionOffset =
+        R.getInstructionOffset() + OffsetCorrection - 2;
+
+    CallSites[RecordIndex] = InstructionOffset;
+    CallSiteSizes[RecordIndex] = 2;
+
+#if !defined(NDEBUG)
+    ostringstream SlotStream;
+    ostringstream LiveStream;
+
+    if (EmitLogs) {
+      LiveStream << "Safepoint: " << RecordIndex << ": ("
+                 << CallSites[RecordIndex] << " - "
+                 << (CallSites[RecordIndex] + CallSiteSizes[RecordIndex])
+                 << ")\n";
+    }
+#endif // !NDEBUG
+
+    // We don't generate GC_CALLER_SP_REL locatons, just
+    // using this as a default value other than SP/FP REL.
+    GcStackSlotBase SpBase = GC_CALLER_SP_REL;
+
+    for (const auto &Loc : R.locations()) {
+
+      switch (Loc.getKind()) {
+      case StackMapParserType::LocationKind::Constant:
+      case StackMapParserType::LocationKind::ConstantIndex:
+        continue;
+
+      case StackMapParserType::LocationKind::Register:
+        // TODO: Report Live - GC values in Registers
+        // https://github.com/dotnet/llilc/issues/474
+        // Live gc-pointers are currently spilled to the stack at Safepoints.
+        assert(false && "GC-Pointer Live in Register");
+        break;
+
+      case StackMapParserType::LocationKind::Direct: {
+
+        uint16_t DwReg = Loc.getDwarfRegNum();
+        switch (DwReg) {
+        case DW_FRAME_POINTER:
+          assert(SpBase != GC_SP_REL && "Mixed SP/FP based Locations");
+          SpBase = GC_FRAMEREG_REL;
+          break;
+        case DW_STACK_POINTER:
+          assert(SpBase != GC_FRAMEREG_REL && "Mixed SP/FP based Locations");
+          SpBase = GC_SP_REL;
+          break;
+        default:
+          assert(false && "Unexpected stack base-pointer");
+        }
+
+        GcSlotId SlotID;
+        int32_t Offset = Loc.getOffset();
+        map<int32_t, GcSlotId>::iterator ExistingSlot = SlotMap.find(Offset);
+        if (ExistingSlot == SlotMap.end()) {
+          SlotID = Encoder->GetStackSlotId(Offset, SlotFlags, SpBase);
```

Comment:
Given the way that these slot IDs are used, an assertion that the newly returned slot id is exactly `NumSlots` would be useful.